### PR TITLE
initialize DB

### PR DESCRIPTION
**Pull request by haianhnc**
Init data base

### CODE REVIEW

File: server/Initialize/generate_table.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+rails g model Article title:string link:string content:text posted_at:datetime author:string
```
Comment (xuandunghp) [inline server/Initialize/generate_table.sh]: server/Initialize/generate_table.sh => server/scripts/generate_models.sh

File: server/app/models/article.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class Article < ApplicationRecord
+  has_and_belongs_to_many :tags
```
Comment (xuandunghp) [inline server/app/models/article.rb]: use `has_many` and `has_many .. through ...`
http://blog.flatironschool.com/why-you-dont-need-has-and-belongs-to-many/

File: server/app/models/article.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+  belongs_to :source
+
```
Comment (xuandunghp) [inline server/app/models/article.rb:11]: Basic validation:
- presence
- length (usually maximum 255 for string field)

File: server/Initialize/generate_table.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+rails g model Tag name:string source_list:string user_id:integer tag:string
```
Comment (xuandunghp) [inline server/Initialize/generate_table.sh]: - string is default, should remove it for conciseness
- user_id:integer => user:references

File: server/config/application-default.yml [hunk 5]
```diff
@@ -1,13 +0,0 @@
-development:
```
Comment (xuandunghp) [inline server/config/application-default.yml:1]: keep this file for future developers, don't delete it

File: server/db/migrate/20161102161630_create_articles.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+class CreateArticles < ActiveRecord::Migration[5.0]
```
Comment (xuandunghp) [inline server/db/migrate/20161102161630_create_articles.rb]: remove [5.0]

Comment (dominhhai610) [reply server/db/migrate/20161102161630_create_articles.rb]: dùng rails generate nó tự có mà anh ơi, bỏ đi bằng tay lần sau generate lại thì nó lại ra đấy

Comment (xuandunghp) [reply server/db/migrate/20161102161630_create_articles.rb]: ok chú, thế thì cứ để tạm thê

File: server/Initialize/generate_table.sh (hunk 4 above)
Comment (dominhhai610) [reply server/Initialize/generate_table.sh]: Mysql2::Error: Cannot add foreign key constraint: CREATE TABLE `tags` (`id` int AUTO_INCREMENT PRIMARY KEY, `name` varchar(255), `source_list` varchar(255), `user_id` int, `tag` varchar(255), `created_at` datetime NOT NULL, `updated_at` datetime NOT NULL,  INDEX `index_tags_on_user_id`  (`user_id`), CONSTRAINT `fk_rails_e689f6d0cc`
FOREIGN KEY (`user_id`)
  REFERENCES `users` (`id`)
) ENGINE=InnoDB

em bị lỗi như này anh ơi

Comment (dominhhai610) [reply server/Initialize/generate_table.sh]: em chạy lại OK ròi nhé, phải tao table User trước mới add FK được.

File: server/app/models/article.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+class Article < ApplicationRecord
+  belongs_to :source
+  has_many :articletag
+  has_many :tags, through: :articletag
```
Comment (xuandunghp) [inline server/app/models/article.rb]: articletag => article_tags

File: server/app/models/article.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+  validates :title, presence: true
+  validates_length_of :title, maximum: 255
```
Comment (xuandunghp) [inline server/app/models/article.rb]: validates :title, presence: true, length: {maximum: 255}

File: server/app/models/article.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  validates :link, presence: true
+  validates_length_of :link, maximum: 255
```
Comment (xuandunghp) [inline server/app/models/article.rb]: validates :link, presence: true, length: {maximum: 255}

File: server/app/models/article.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  validates :content, presence :true
```
Comment (xuandunghp) [inline server/app/models/article.rb]: Always have a maximum length for text field
=> validates :content, presence :true, length: {maximum: 10000}

File: server/app/models/articletag.rb [hunk 11]
```diff
@@ -0,0 +1,4 @@
+class Articletag < ApplicationRecord
```
Comment (xuandunghp) [inline server/app/models/articletag.rb]: Articletag => ArticleTag

File: server/app/models/articletag.rb [hunk 12]
```diff
@@ -0,0 +1,4 @@
...
+  belongs_to :article
+  belongs_to :tags
```
Comment (xuandunghp) [inline server/app/models/articletag.rb]: belongs_to :tags => belongs_to :tag

File: server/app/models/liked_article.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class LikedArticle < ApplicationRecord
+  belongs_to :user
+
+  validates :article_id, presence: true
```
Comment (xuandunghp) [inline server/app/models/liked_article.rb]: - remove validates
- add belongs_to :article

File: server/app/models/source.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+class Source < ApplicationRecord
+  belongs_to :tag
```
Comment (xuandunghp) [inline server/app/models/source.rb]: remove. There is no relation between Source and Tag

File: server/app/models/tag.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+class Tag < ApplicationRecord
+  has_many :sources
```
Comment (xuandunghp) [inline server/app/models/tag.rb]: remove

File: server/app/models/user.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  has_many :LikedArticles
+  has_many :User_customize_sources
+  has_many :User_customize_tags
```
Comment (xuandunghp) [inline server/app/models/user.rb]: remove these 2 lines. Not needed for now

File: server/app/models/user_customize_source.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class UserCustomizeSource < ApplicationRecord
```
Comment (xuandunghp) [inline server/app/models/user_customize_source.rb]: remove this model

File: server/app/models/user_customize_tag.rb [hunk 18]
```diff
@@ -0,0 +1,6 @@
+class UserCustomizeTag < ApplicationRecord
```
Comment (xuandunghp) [inline server/app/models/user_customize_tag.rb]: remove this model

File: server/db/migrate/20161102175404_create_user_customize_tags.rb [hunk 19]
```diff
@@ -0,0 +1,10 @@
+class CreateUserCustomizeTags < ActiveRecord::Migration[5.0]
```
Comment (xuandunghp) [inline server/db/migrate/20161102175404_create_user_customize_tags.rb]: remove this migration

File: server/db/migrate/20161103093159_create_users.rb [hunk 20]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :user_name
+      t.string :password
+      t.integer :is_admin
```
Comment (xuandunghp) [inline server/db/migrate/20161103093159_create_users.rb]: use a different model for admin

File: server/db/migrate/20161103093704_create_liked_articles.rb [hunk 21]
```diff
@@ -0,0 +1,10 @@
+class CreateLikedArticles < ActiveRecord::Migration[5.0]
+  def change
+    create_table :liked_articles do |t|
+      t.references :user, foreign_key: true
+      t.integer :article_id
```
Comment (xuandunghp) [inline server/db/migrate/20161103093704_create_liked_articles.rb]: t.references :article, foreign_key: true

File: server/db/migrate/20161103093632_create_tags.rb [hunk 22]
```diff
@@ -0,0 +1,12 @@
+class CreateTags < ActiveRecord::Migration[5.0]
+  def change
+    create_table :tags do |t|
+      t.string :name
+      t.string :source_list
+      t.references :user, foreign_key: true
+      t.string :tag
```
Comment (xuandunghp) [inline server/db/migrate/20161103093632_create_tags.rb]: don't need this field. Use `name` for tag content.

File: server/db/migrate/20161103093737_create_user_customize_sources.rb [hunk 23]
```diff
@@ -0,0 +1,10 @@
+class CreateUserCustomizeSources < ActiveRecord::Migration[5.0]
```
Comment (xuandunghp) [inline server/db/migrate/20161103093737_create_user_customize_sources.rb]: remove this migration

File: server/db/migrate/20161103093825_create_user_customize_tags.rb [hunk 24]
```diff
@@ -0,0 +1,10 @@
+class CreateUserCustomizeTags < ActiveRecord::Migration[5.0]
```
Comment (xuandunghp) [inline server/db/migrate/20161103093825_create_user_customize_tags.rb]: remove this migration

File: server/scripts/generate_models.sh [hunk 25]
```diff
@@ -0,0 +1,7 @@
+rails g model User user_name password is_admin:integer
```
Comment (xuandunghp) [inline server/scripts/generate_models.sh]: model names are usually in snake case or camel case, for example:
- `rails g model liked_article user:references article_id:integer`
  OR
- `rails g model LikedArticle user:references article_id:integer`

File: server/app/models/article.rb (hunk 7 above)
Comment (haianhnc) [reply server/app/models/article.rb]: fixed

**Pull request merged**